cou/steps/plan.py

```diff
@@ -416,58 +429,105 @@ def _generate_control_plane_plan(
     :rtype: list[UpgradePlan]
     """
     principal_upgrade_plan = create_upgrade_group(
-        apps=apps,
+        apps=[app for app in apps if app.is_subordinate is False],
         description="Control Plane principal(s) upgrade plan",
         target=target,
         force=force,
-        filter_function=lambda app: not isinstance(app, SubordinateBase),
     )
 
     subordinate_upgrade_plan = create_upgrade_group(
-        apps=apps,
+        apps=[app for app in apps if app.is_subordinate],
         description="Control Plane subordinate(s) upgrade plan",
         target=target,
         force=force,
-        filter_function=lambda app: isinstance(app, SubordinateBase),
     )
 
     logger.debug("Generation of the control plane upgrade plan complete")
     return [principal_upgrade_plan, subordinate_upgrade_plan]
 
 
-async def _generate_data_plane_plan(
-    target: OpenStackRelease,
-    analysis_result: Analysis,
+def _separate_hypervisors_apps(
+    apps_data_plane: list[OpenStackApplication],
+) -> tuple[list[OpenStackApplication], list[OpenStackApplication]]:
+    """Separate what is considered hypervisors apps from non-hypervisors apps.
+
+    :param apps_data_plane: Applications from data-plane
+    :type apps_data_plane: list[OpenStackApplication]
+    :raises DataPlaneCannotUpgrade: When an unknown data-plane app is passed.
+    :return: Tuple containing two lists of hypervisors and non-hypervisors apps
+    :rtype: tuple[list[OpenStackApplication], list[OpenStackApplication]]
+    """
+    hypervisor_apps = []
+    non_hypervisors_apps = []
+    _, nova_compute_machines = _get_nova_compute_units_and_machines(apps_data_plane)
+    for app in apps_data_plane:
+        if app.charm == "ceph-osd" or app.is_subordinate:
+            non_hypervisors_apps.append(app)
+        elif any(unit.machine in nova_compute_machines for unit in app.units.values()):
+            hypervisor_apps.append(app)
+        else:
+            raise DataPlaneCannotUpgrade(f"COU does not know how to upgrade '{app.name}'")
```

Comment:
Is this really needed if we already have a supported list of data-plane apps defined? It's nice that we are being prudent, but I think it's also important to prune unnecessary code.

Reply:
I agree with removing this.